standards/docker_standards.md

```diff
@@ -0,0 +1,230 @@
+
+A comparison of the different tools is available in [Confluence](https://eaflood.atlassian.net/wiki/spaces/FPS/pages/1408172407/Container+Image+Scanning+Tools)
+
+## Multi stage builds
+Dockerfiles should implement multi stage builds to allow different container configuration and behaviour to be created for development, test and production.
+
+Below is an example multi stage build which is intended to use the Future Farming and Countryside (FFC) Node.js parent image.
+
+```
+# Base stage installs production dependencies
+ARG REGISTRY=562955126301.dkr.ecr.eu-west-2.amazonaws.com
+ARG BASE_VERSION=1.0.0
+ARG DEV_VERSION=1.0.0
+FROM $REGISTRY/ffc-node-base:$BASE_VERSION as base
+ARG PORT=3000
+ENV PORT ${PORT}
+USER node
+WORKDIR /home/node
+COPY --chown=node:node package*.json ./
+RUN npm ci
+
+# Development stage installs devDependencies, builds app from source and declares a file watcher as the default command
+FROM $REGISTRY/ffc-node-development:$DEV_VERSION AS development
+EXPOSE ${PORT} 9229 9230
+USER node
+WORKDIR /home/node
+COPY --from=base --chown=node:node /home/node/package*.json ./
+COPY --from=base --chown=node:node /home/node/node_modules ./node_modules
+RUN npm install
+COPY --chown=node:node app/ ./app/
+RUN npm run build
+CMD [ "npm", "run", "start:watch" ]
+
+# Test stage copied in Jest configuration and declares the test task as the default command
+FROM development AS test
+USER node
+WORKDIR /home/node
+COPY --chown=node:node jest.config.js ./jest.config.js
+COPY --chown=node:node test/ ./test/
+CMD [ "npm", "run", "test" ]
+
+# Production stage exposes service port, copies in built app code and declares the Node app as the default command
+FROM base AS production
+USER node
+WORKDIR /home/node
+EXPOSE ${PORT}
+COPY --from=development /home/node/app/ ./app/
+CMD [ "node", "app/index" ]
+```
+
+## Docker Compose
+Docker compose can be used for local development and within a CI/CD pipeline to build docker containers.
+
+### Use override files to reduce duplication
+Additional settings can be applied to a docker compose file by using override files.
+
+Override files can be applied by listing the files after the `docker-compose` command with the `-f` parameter, i.e.
+
+`docker-compose up -f docker-compose.yaml -f docker-compose.override.yaml`
+
+Note that the above is equivalent to running the command:
+
+`docker-compose up`
+
+as calling `docker-compose` without specifying any files will run `docker-compose` with any available `docker-compose.yaml` and `docker-compose.override.yaml` files in the executing directory. 
+
+Note however that:
+
+`docker-compose up -f docker-compose.yaml`
+
+will **not** apply the docker `docker-compose.override.yaml` file, only the file specified.
+
+One use case is for testing - common settings can be put into the base `docker-compose.yaml` file, while changes to the command and container names can be placed in override files.
```

Comment:
Should this be

> One use case is for ~testing~ running tests in CI

?

I'd expect to be running unit tests locally on my dev container using a locally-mounted volume of my source.

Reply:
Why would you not run unit tests locally?  I regularly run unit tests locally.

Reply:
(deleted my comment as I misread yours)

Agreed.

Reply:
Updated.